src/main/scala/analysis/BasicIRConstProp.scala

```diff
@@ -0,0 +1,91 @@
+package analysis
+import ir.*
+import analysis.solvers.*
+
+trait ILValueAnalysisMisc:
+  val valuelattice: ConstantPropagationLattice = ConstantPropagationLattice()
+  val statelattice: MapLattice[Variable, FlatElement[BitVecLiteral], ConstantPropagationLattice] = MapLattice(valuelattice)
+
+  def eval(exp: Expr, env: statelattice.Element): valuelattice.Element =
+    import valuelattice._
+    exp match
+      case id: Variable   => env(id)
+      case n: BitVecLiteral     => bv(n)
+      case ze: ZeroExtend => zero_extend(ze.extension, eval(ze.body, env))
+      case se: SignExtend => sign_extend(se.extension, eval(se.body, env))
+      case e: Extract     => extract(e.end, e.start, eval(e.body, env))
+      case bin: BinaryExpr =>
+        val left = eval(bin.arg1, env)
+        val right = eval(bin.arg2, env)
+        bin.op match
+          case BVADD  => bvadd(left, right)
+          case BVSUB  => bvsub(left, right)
+          case BVMUL  => bvmul(left, right)
+          case BVUDIV => bvudiv(left, right)
+          case BVSDIV => bvsdiv(left, right)
+          case BVSREM => bvsrem(left, right)
+          case BVUREM => bvurem(left, right)
+          case BVSMOD => bvsmod(left, right)
+          case BVAND  => bvand(left, right)
+          case BVOR   => bvor(left, right)
+          case BVXOR  => bvxor(left, right)
+          case BVNAND => bvnand(left, right)
+          case BVNOR  => bvnor(left, right)
+          case BVXNOR => bvxnor(left, right)
+          case BVSHL  => bvshl(left, right)
+          case BVLSHR => bvlshr(left, right)
+          case BVASHR => bvashr(left, right)
+          case BVCOMP => bvcomp(left, right)
+          case BVCONCAT => concat(left, right)
+
+          //case BVULE => bvule(left, right)
+          //case BVUGE => bvuge(left, right)
+          //case BVULT => bvult(left, right)
+          //case BVUGT => bvugt(left, right)
+
+          //case BVSLE => bvsle(left, right)
+          //case BVSGE => bvsge(left, right)
+          //case BVSLT => bvslt(left, right)
+          //case BVSGT => bvsgt(left, right)
+
+          //case BVCONCAT => concat(left, right)
+          //case BVNEQ    => bvneq(left, right)
+          //case BVEQ     => bveq(left, right)
+
+      case un: UnaryExpr =>
+        val arg = eval(un.arg, env)
+
+        un.op match
+          case BVNOT => bvnot(arg)
+          case BVNEG => bvneg(arg)
+
+      case _ => valuelattice.top
+
+  val calleePreservedRegisters = Set("R0", "R1", "R2", "R3", "R4", "R5", "R6", "R7", "R8", "R9", "R10", "R11")
```

Comment:
These are not the callee-preserved registers, which are the ones that must be preserved by a subroutine call. Those are R19-R29 and R31.

What this should actually be called is the caller-preserved registers (which means it's the caller's responsibility to preserve them if it wants to) and it should consist of R0 to R18, and R30.